File: java/4-LongestCommonPrefix.java
// Horizontal scanning

class LongestCommonPrefix {
    public String longestCommonPrefix(String[] strs) {
        // Edge case: If the input array is null or empty, return an empty string
        if (strs == null || strs.length == 0) {
            return "";
        }

        // Initialize the prefix as the entire first string
        String prefix = strs[0];

        // Compare the prefix with all other strings in the array
        for (int i = 1; i < strs.length; i++) {
            // While the current string strs[i] does not start with the prefix
            while (strs[i].indexOf(prefix) != 0) {
                // Shorten the prefix by removing its last character
                prefix = prefix.substring(0, prefix.length() - 1);

                // If the prefix becomes empty, there is no common prefix
                if (prefix.isEmpty()) {
                    return "";
                }
            }
        }

        return prefix;
    }
}

// Logic Explanation

// The horizontal scanning strategy slices through the problem by narrowing down
// the answer string-by-string.

// 1. Step-by-Step BreakdownImagine the input is ["flower", "flow", "flight"].
// - Initialization: We set prefix = "flower".
// - First Comparison (with "flow"): We check if "flow" starts with "flower"
// using strs[i].indexOf(prefix). It doesn't (returns -1), so we cut "flower"
// down to "flowe", then down to "flow". Now, "flow".indexOf("flow") == 0. The
// loop conditions are satisfied, and our updated prefix is "flow".
// - Second Comparison (with "flight"): We check if "flight" starts with "flow".
// It doesn't.
// -- "flow" shortens to "flo" $\rightarrow$ still no match.
// -- "flo" shortens to "fl" $\rightarrow$ Match found! "flight" starts with
// "fl" at index 0.
// Termination: We have exhausted the array, leaving "fl" as the final longest
// common prefix.

// 2. Early Exit GuardIf at any point during the truncation the prefix is
// chopped down to an empty string "", the loop catches it immediately via
// prefix.isEmpty() and returns "". This saves precious iterations if the first
// two words share nothing in common (e.g., ["dog", "racecar"]).

File: python/easy/4_longest_common_prefix.py
# Vertical Scanning

from typing import List

class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if not strs:
            return ""
        
        # Take the first string as the baseline reference
        for i, char in enumerate(strs[0]):
            # Check this character against all other strings
            for string in strs[1:]:
                # If we exceed the bounds of the current string or find a mismatch
                if i == len(string) or string[i] != char:
                    return strs[0][:i]
                    
        return strs[0]
    
# Step-by-Step Execution
# Edge Case: If the input list strs is empty, return an empty string "" immediately.

# Take the first string strs[0] as the reference string.

# Iterate through each character of the reference string using its index i and character char.

# For every character, loop through the rest of the strings in the list (from index 1 to the end):

# Check if the current index i is equal to the length of the string (meaning we ran out of characters in this string), OR if the character at string[i] does not match char.

# If either condition is true, we have found the end of the common prefix. Return the substring from the reference string from the beginning up to index i (strs[0][:i]).

# If the loops finish without any mismatch, it means the entire reference string (or the shortest string in the list) is the common prefix. Return the baseline string.    